File: lib/gherkin/i18n.rb
require 'yaml'
require 'gherkin/rubify'
require 'gherkin/bridge'

module Gherkin
  class I18n
    native_impl('gherkin') unless defined?(BYPASS_NATIVE_IMPL)

    FEATURE_ELEMENT_KEYS = %w{feature background scenario scenario_outline examples}
    STEP_KEYWORD_KEYS    = %w{given when then and but}
    KEYWORD_KEYS         = FEATURE_ELEMENT_KEYS + STEP_KEYWORD_KEYS
    LANGUAGES            = YAML.load_file(File.dirname(__FILE__) + '/i18n.yml')

    class << self
      include Rubify

      # Used by code generators for other lexer tools like pygments lexer and textmate bundle
      def all
        LANGUAGES.keys.sort.map{|iso_code| get(iso_code)}
      end

      def get(iso_code)
        languages[iso_code] ||= new(iso_code)
      end

      # Returns all keyword translations and aliases of +keywords+, escaped and joined with <tt>|</tt>.
      # This method is convenient for editor support and syntax highlighting engines for Gherkin, where
      # there is typically a code generation tool to generate regular expressions for recognising the
      # various I18n translations of Gherkin's keywords.
      #
      # The +keywords+ arguments can be one of <tt>:feature</tt>, <tt>:background</tt>, <tt>:scenario</tt>, 
      # <tt>:scenario_outline</tt>, <tt>:examples</tt>, <tt>:step</tt>.
      def keyword_regexp(*keywords)
        unique_keywords = all.map do |i18n|
          keywords.map do |keyword|
            if keyword.to_s == 'step'
              i18n.step_keywords
            else
              i18n.keywords(keyword)
            end
          end
        end
        
        unique_keywords.flatten.compact.sort.reverse.uniq.join('|').gsub(/\*/, '\*')
      end

      def code_keywords
        rubify(all.map{|i18n| i18n.code_keywords}).flatten.uniq.sort
      end

      def code_keyword_for(gherkin_keyword)
        gherkin_keyword.gsub(/[\s',]/, '').strip
      end

      def language_table
        require 'stringio'
        require 'gherkin/formatter/pretty_formatter'
        io = defined?(JRUBY_VERSION) ? Java.java.io.StringWriter.new : StringIO.new
        pf = Gherkin::Formatter::PrettyFormatter.new(io, true)
        all.each{|i18n| pf.row([i18n.iso_code, i18n.keywords('name')[0], i18n.keywords('native')[0]], 0)}
        pf.flush_table
        if defined?(JRUBY_VERSION)
          io.getBuffer.toString
        else
          io.rewind
          io.read
        end
      end

      def unicode_escape(word, prefix="\\u")
        word = word.unpack("U*").map do |c|
          if c > 127 || c == 32
            "#{prefix}%04x" % c
          else
            c.chr
          end
        end.join
      end

      private

      def languages
        @languages ||= {}
      end
    end

    attr_reader :iso_code

    def initialize(iso_code)
      @iso_code = iso_code
      @keywords = LANGUAGES[iso_code]
      raise "Language not supported: #{iso_code.inspect}" if @iso_code.nil?
      @keywords['grammar_name'] = @keywords['name'].gsub(/\s/, '')
    end

    def lexer(listener, force_ruby=false)
      begin
        if force_ruby
          rb(listener)
        else
          begin
            c(listener)
          rescue NameError, LoadError => e
            warn("WARNING: #{e.message}. Reverting to Ruby lexer.")
            rb(listener)
          end
        end
      rescue LoadError => e
        raise I18nLexerNotFound, "No lexer was found for #{i18n_language_name} (#{e.message}). Supported languages are listed in gherkin/i18n.yml."
      end
    end

    def c(listener)
      require 'gherkin/c_lexer'
      CLexer[underscored_iso_code].new(listener)
    end

    def rb(listener)
      require 'gherkin/rb_lexer'
      RbLexer[underscored_iso_code].new(listener)
    end

    def underscored_iso_code
      @iso_code.gsub(/[\s-]/, '_').downcase
    end

    # Keywords that can be used in Gherkin source
    def step_keywords
      STEP_KEYWORD_KEYS.map{|iso_code| keywords(iso_code)}.flatten.uniq
    end

    # Keywords that can be used in code
    def code_keywords
      result = step_keywords.map{|keyword| self.class.code_keyword_for(keyword)}
      result.delete('*')
      result
    end

    def keywords(iso_code)
      iso_code = iso_code.to_s
      raise "No #{iso_code.inspect} in #{@keywords.inspect}" if @keywords[iso_code].nil?
      @keywords[iso_code].split('|').map{|keyword| real_keyword(iso_code, keyword)}
    end

    def keyword_table
      require 'stringio'
      require 'gherkin/formatter/pretty_formatter'
      io = StringIO.new
      pf = Gherkin::Formatter::PrettyFormatter.new(io, true)

      KEYWORD_KEYS.each do |key|
        pf.row([key, keywords(key).map{|keyword| %{"#{keyword}"}}.join(', ')], 0)
      end
      STEP_KEYWORD_KEYS.each do |key|
        code_keywords = keywords(key).reject{|keyword| keyword == '* '}.map do |keyword|
          %{"#{self.class.code_keyword_for(keyword)}"}
        end.join(', ')
        pf.row(["#{key} (code)", code_keywords], 0)
      end
      
      pf.flush_table
      io.rewind
      io.read
    end

    private

    def real_keyword(iso_code, keyword)
      if(STEP_KEYWORD_KEYS.index(iso_code))
        (keyword + ' ').sub(/< $/, '')
      else
        keyword
      end
    end
  end
end

File: python/gherkin/i18n.py
# -*- coding: utf8 -*-

import os.path
import re

import yaml

here = os.path.dirname(__file__)
code_keyword_re = re.compile(r"[\s',!]")
whitespace_re = re.compile(r"\s")
underscore_re = re.compile(r"[\s-]")
real_keyword_re = re.compile(r"< $")

class I18n(object):
    FEATURE_ELEMENT_KEYS = [
        'feature',
        'background',
        'scenario',
        'scenario_outline',
        'examples',
    ]
    STEP_KEYWORD_KEYS = [
        'given',
        'when',
        'then',
        'and',
        'but',
    ]
    KEYWORD_KEYS = FEATURE_ELEMENT_KEYS + STEP_KEYWORD_KEYS
    LANGUAGES = yaml.load(open(os.path.join(here, 'i18n.yml')))

    languages = {}

    @classmethod
    def all(cls):
        codes = cls.LANGUAGES.keys()
        codes.sort()
        return [cls.get(iso_code) for iso_code in codes]

    @classmethod
    def get(cls, iso_code):
        i18n = cls.languages.get(iso_code, None)
        if i18n is not None:
            return i18n
        i18n = cls.languages[iso_code] = I18n(iso_code)
        return i18n

    @classmethod
    def keyword_regexp(cls, *keywords):
        unique_keywords = set()
        for i18n in cls.all():
            for keyword in keywords:
                if str(keyword) == 'step':
                    unique_keywords.update(i18n.step_keywords())
                else:
                    unique_keywords.update(i18n.keywords(keyword))

        unique_keywords = [unicode(k) for k in unique_keywords]
        unique_keywords.sort(reverse=True)
        return '|'.join(unique_keywords)

    @classmethod
    def all_code_keywords(cls):
        keywords = set(sum([i18n.code_keywords() for i18n in cls.all()], []))
        keywords = list(keywords)
        keywords.sort()
        return keywords

    @classmethod
    def code_keyword_for(cls, gherkin_keyword):
        return code_keyword_re.sub('', gherkin_keyword).strip()

    @classmethod
    def unicode_escape(cls, word, prefix="\\u"):
        chars = []
        for c in [ord(char) for char in word]:
            if c > 127 or c == 32:
                chars.append("%s%04x" % (prefix, c))
            else:
                chars.append(unichr(c))

        return ''.join(chars)

    def __init__(self, iso_code):
        self.iso_code = iso_code
        self._keywords = self.LANGUAGES.get(iso_code, None)
        if self._keywords is None:
            raise Exception("Language not supported: %r" % iso_code)
        self._keywords['grammar_name'] = \
            whitespace_re.sub('', self._keywords['name'])
        self.underscored_iso_code = \
            underscore_re.sub('_', self.iso_code).lower()

    def lexer(self, listener):
        pass

    def step_keywords(self):
        keywords = set()
        for iso_code in self.STEP_KEYWORD_KEYS:
            keywords.update(self.keywords(iso_code))
        return list(keywords)

    def code_keywords(self):
        keywords = [self.code_keyword_for(k) for k in self.step_keywords()]
        return [k for k in keywords if k != '*']

    def keywords(self, key):
        key = str(key)
        keywords = self._keywords[key].split('|')
        return [self.real_keyword(key, keyword) for keyword in keywords]

    def real_keyword(self, key, keyword):
        if key in self.STEP_KEYWORD_KEYS:
            return real_keyword_re.sub('', keyword + ' ')
        else:
            return keyword